### ISSUE TITLE
Error building libsupc++ (part of GCC 6 build)

### ISSUE THREAD

**Issue by sebastianbergmann** [bebbo/amigaos-cross-toolchain#5]
```
Making all in libsupc++
make[2]: Entering directory '/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/libsupc++'
/bin/bash ../libtool --tag CXX --tag disable-shared   --mode=compile /root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/./gcc/xgcc -shared-libgcc -B/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/./gcc -nostdinc++ -L/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/src -L/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/src/.libs -L/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/libsupc++/.libs -B/opt/m68k-amigaos/m68k-amigaos/bin/ -B/opt/m68k-amigaos/m68k-amigaos/lib/ -isystem /opt/m68k-amigaos/m68k-amigaos/include -isystem /opt/m68k-amigaos/m68k-amigaos/sys-include    -I/root/amigaos-cross-toolchain/submodules/gcc-6/libstdc++-v3/../libgcc -I/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/include/m68k-amigaos -I/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/include -I/root/amigaos-cross-toolchain/submodules/gcc-6/libstdc++-v3/libsupc++   -prefer-pic -D_GLIBCXX_SHARED -fno-implicit-templates -Wall -Wextra -Wwrite-strings -Wcast-qual -Wabi  -fdiagnostics-show-location=once      -frandom-seed=vterminate.lo -g -O2  -c -o vterminate.lo /root/amigaos-cross-toolchain/submodules/gcc-6/libstdc++-v3/libsupc++/vterminate.cc
libtool: compile:  /root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/./gcc/xgcc -shared-libgcc -B/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/./gcc -nostdinc++ -L/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/src -L/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/src/.libs -L/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/libsupc++/.libs -B/opt/m68k-amigaos/m68k-amigaos/bin/ -B/opt/m68k-amigaos/m68k-amigaos/lib/ -isystem /opt/m68k-amigaos/m68k-amigaos/include -isystem /opt/m68k-amigaos/m68k-amigaos/sys-include -I/root/amigaos-cross-toolchain/submodules/gcc-6/libstdc++-v3/../libgcc -I/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/include/m68k-amigaos -I/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/include -I/root/amigaos-cross-toolchain/submodules/gcc-6/libstdc++-v3/libsupc++ -D_GLIBCXX_SHARED -fno-implicit-templates -Wall -Wextra -Wwrite-strings -Wcast-qual -Wabi -fdiagnostics-show-location=once -frandom-seed=vterminate.lo -g -O2 -c /root/amigaos-cross-toolchain/submodules/gcc-6/libstdc++-v3/libsupc++/vterminate.cc  -D_GLIBCXX_SHARED -o vterminate.o
In file included from /root/amigaos-cross-toolchain/submodules/gcc-6/libstdc++-v3/libsupc++/vterminate.cc:32:0:
/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/include/cstdio:176:11: error: '::vfscanf' has not been declared
   using ::vfscanf;
           ^~~~~~~
/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/include/cstdio:186:22: error: '__gnu_cxx::vfscanf' has not been declared
   using ::__gnu_cxx::vfscanf;
                      ^~~~~~~
Makefile:653: recipe for target 'vterminate.lo' failed
make[2]: *** [vterminate.lo] Error 1
make[2]: Leaving directory '/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3/libsupc++'
Makefile:507: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/root/amigaos-cross-toolchain/.build-m68k/build/gcc-6/m68k-amigaos/libstdc++-v3'
Makefile:414: recipe for target 'all' failed
make: *** [all] Error 2
```

Below is the `Dockerfile` I use to build the toolchain:
```Dockerfile
# Ubuntu 16.04 ("Xenial Xerus") 32-bit
FROM daald/ubuntu32:xenial

# Update base system and install build dependencies
RUN apt-get update && apt-get install -y \
    autoconf \
    bison \
    g++ \
    gcc \
    gettext \
    git \
    gperf \
    libgmp-dev \
    libmpc-dev \
    libmpfr-dev \
    libncurses5-dev \
    make \
    python-dev \
    && rm -rf /var/lib/apt/lists/*

# Build M68K AmigaOS Cross-Compilation Toolchain
RUN cd /root && \
    git clone https://github.com/bebbo/amigaos-cross-toolchain.git && \
    cd /root/amigaos-cross-toolchain && \
    ./toolchain-m68k --prefix=/opt/m68k-amigaos build
```

**Comment by MBeijer** [comment]
I believe bebbo is aware of this, since it is a work in progress and not
fully working yet.
You can follow some discussion here:
http://eab.abime.net/showthread.php?t=85474

**Comment by sebastianbergmann** [comment]
Thank you, @MBeijer, for pointing me to that thread on EAB.

Correct me if I am wrong, but the C compiler seems to be working and we're stuck on the C++ compiler (and/or its libraries). If that is correct, would it not make sense, @bebbo, to only build with `--enable-languages=c` for now in the `master` branch and maybe have a working build? A branch could then be used to work on getting C++ to work.

Thoughts?

**Comment by bebbo** [comment]
The build will still fail.

Feel free to suggest usefule fixes!

**Comment by chenz** [comment]
I think the problem is that the configuration detects a C99/C11-compliant stdio.h, when in fact it should not? In libnix for example, `vfscanf` is a macro. The checks in acinclude.m4 succeed with the macros, but then libstdc++'s cstdio undefs them, expecting them to be provided as functions instead, leading to the compilation error above. I hacked around this, but did not get much further after that.

```
diff --git a/libstdc++-v3/acinclude.m4 b/libstdc++-v3/acinclude.m4
index ce44a6a..ceb6296 100644
--- a/libstdc++-v3/acinclude.m4
+++ b/libstdc++-v3/acinclude.m4
@@ -1041,6 +1041,11 @@ AC_DEFUN([GLIBCXX_ENABLE_C99], [
       GCC_TRY_COMPILE_OR_LINK(
         [#include <stdio.h>
          #include <stdarg.h>
+         #undef snprintf
+         #undef vfscanf
+         #undef vscanf
+         #undef vsnprintf
+         #undef vsscanf
          void foo(char* fmt, ...)
          {
            va_list args; va_start(args, fmt);
@@ -1278,6 +1283,11 @@ AC_DEFUN([GLIBCXX_ENABLE_C99], [
       GCC_TRY_COMPILE_OR_LINK(
         [#include <stdio.h>
          #include <stdarg.h>
+         #undef snprintf
+         #undef vfscanf
+         #undef vscanf
+         #undef vsnprintf
+         #undef vsscanf
          void foo(char* fmt, ...)
          {
            va_list args; va_start(args, fmt);
```

**Comment by sebastianbergmann** [comment]
Fixed by af4c46a5f39c20b78ca408bfc36b48b922b4da97.
